presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -825,6 +825,15 @@ public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Expression> cont
             postJoinConjuncts.addAll(equalityPartition.getScopeComplementEqualities());
             postJoinConjuncts.addAll(equalityPartition.getScopeStraddlingEqualities());
 
+            // Add source predicates inferred from equality on join symbols and filter side predicate
+            Expression filteringSourceEffectivePredicate = EffectivePredicateExtractor.extract(rewrittenFilteringSource);
+            EqualityInference joinSymbolFilteringSourceInference = createEqualityInference(
+                    filteringSourceEffectivePredicate,
+                    new ComparisonExpression(ComparisonExpressionType.EQUAL, node.getSourceJoinSymbol().toSymbolReference(), node.getFilteringSourceJoinSymbol().toSymbolReference()));
```

Comment:
Most likely this is not valid rewrite (correct me if I am wrong). This is because you don't know how semi join mask symbol will be used and it might be used not for filtering but in projection. If it's used in projection then you cannot filter any rows.

Why this rewrite is needed?

Reply:
You might be right. I will give it a thought. 
It is for supporting predicate pushdown for queries like one in test.

Reply:
> This is because you don't know how semi join mask symbol will be used and it might be used not for filtering but in projection. If it's used in projection then you cannot filter any rows.

I don't think that matters. Logically, the rows from the probe side of a semijoin need to satisfy any predicates over the output columns of subquery side. Whatever happens after the semijoin (projections, etc) would only see the filtered rows anyway.

Reply:
> Logically, the rows from the probe side of a semijoin need to satisfy any predicates over the output columns of subquery side

That is not true given what `SemiJoinNode` in Presto means. `SemiJoinNode` is currently used for implementing both `IN (subquery)` and `NOT IN (subquery)` (while in terms of algebra the latter should be implemented using `AntiJoinNode` which is not present right now). This alone makes this PR invalid.

Reply:
Yup, I was responding to the comment about the mask symbol being used for projections.

Reply:
> Yeah - you are right. We cannot assume equality. 

In `PredicatePushDown`, don't you have information that symbol produced by `SemiJoin` must be true (in the cases where it is so)?

Reply:
> In PredicatePushDown, don't you have information that symbol produced by SemiJoin must be true (in the cases where it is so)?

Yeah - I came into conclusion to see if SemiJoin.outputSymbol is part of `inhertiedPredicate` in non-negated form and to perform the rewrite only then.
I did not implement it yet though.

Reply:
>  Logically, the rows from the probe side of a semijoin need to satisfy any predicates over the output columns of subquery side. Whatever happens after the semijoin (projections, etc) would only see the filtered rows anyway.

@martint what do you mean? what about:
```
select x in (...) from ... t(x)
```
this will be translated into `projection` above `semijoin` without any filter. I don't see how would you filter any rows in this case?

Reply:
Ah, yes, you're absolutely right. I forgot about that usage and was thinking of SemiJoin as doing the filtering directly, but that's not how it works.

I guess the way to solve this is to match Filter(SemiJoin(...)) where the the filter evaluates the output of semijoin result and rewrite it based on the effective predicate of the subquery side. It could be implemented as a new independent Rule, not necessarily as part of PredicatePushdown.